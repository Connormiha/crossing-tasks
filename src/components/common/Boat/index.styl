@keyframes shaker
    10%, 90%
        transform: translate(-4px, 0)

    20%, 80%
        transform: translate(4px, 0)

    30%, 50%, 70%
        transform: translate(-6px, 0)

    40%, 60%
        transform: translate(6px, 0)

.boat
    position: absolute
    background: url('assets/images/boat.svg') no-repeat
    background-size: contain
    height: 70px
    width: 50px
    transition: all 1s

    &_position
        &_left
            top: 154px
            left: 246px

        &_right
            top: 154px
            left: 518px

    &_invalid
        animation: shaker 0.5s

    &__character
        margin: 0 0.25rem
