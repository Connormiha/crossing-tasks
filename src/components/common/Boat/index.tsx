import style from './index.styl';

import React from 'react';
import Сharacter from 'components/common/Character';
import { ICharacterBase } from 'flux/types';

import { PositionCharacter } from 'games';

import bem from 'bem-css-modules';

const b = bem(style);

interface IProps extends React.Props<any> {
  items: string[];
  characters: ICharacterBase<string>[];
  position: PositionCharacter;
  invalid: boolean;
  onMoveCharacter(id: string): void;
  onMoveEnd(): void;
  onShakeEnd(): void;
}

export default class Boat extends React.PureComponent<IProps> {
  constructor(props: IProps) {
    super(props);

    this.handleAnimationEnd = this.handleAnimationEnd.bind(this);
  }

  renderItems() {
    const { items, characters, onMoveCharacter } = this.props;

    return items.map((id: string) => {
      return (
        <div className={b('character')} key={id}>
          <Сharacter
            name={characters.find((item) => item.id === id)!.name}
            id={id}
            onClick={onMoveCharacter}
            packed
          />
        </div>
      );
    });
  }

  handleAnimationEnd() {
    if (this.props.invalid) {
      this.props.onShakeEnd();
    }
  }

  render() {
    const { position, invalid, onMoveEnd } = this.props;

    return (
      <div
        className={b({ position, invalid })}
        onTransitionEnd={onMoveEnd}
        onAnimationEnd={this.handleAnimationEnd}
      >
        {this.renderItems()}
      </div>
    );
  }
}
