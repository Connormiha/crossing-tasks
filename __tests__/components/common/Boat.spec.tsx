import * as renderer from 'react-test-renderer';
import Boat from 'components/common/Boat';

import noop from 'lodash/noop';

/* tslint:disable:jsx-wrap-multiline */
describe('<Boat />', () => {
    it('should render', () => {
        let tree = renderer.create(<Boat
            items={[]}
            position="left"
            onMoveEnd={noop}
            onMoveCharacter={noop}
            invalid={false}
        />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render items', () => {
        let items = [0, 1],
            characters = {0: {name: 'sheep'}, 1: {name: 'farmer'}},
            tree = renderer.create(
                <Boat
                    items={items}
                    characters={characters}
                    position="left"
                    invalid={false}
                    onMoveEnd={noop}
                    onMoveCharacter={noop}
                />
            ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
