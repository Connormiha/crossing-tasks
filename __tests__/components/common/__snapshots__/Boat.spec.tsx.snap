// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Boat /> should render 1`] = `
<div
  className="boat boat_position_left"
  onTransitionEnd={[Function]}
>
  <audio
    src="assets/audio/wave.mp3"
  />
</div>
`;

exports[`<Boat /> should render items 1`] = `
<div
  className="boat boat_position_left"
  onTransitionEnd={[Function]}
>
  <div
    className="boat__character"
  >
    <div
      className="character character_skin_sheep character_packed"
      onClick={[Function]}
      title="sheep"
    />
  </div>
  <div
    className="boat__character"
  >
    <div
      className="character character_skin_farmer character_packed"
      onClick={[Function]}
      title="farmer"
    />
  </div>
  <audio
    src="assets/audio/wave.mp3"
  />
</div>
`;
